cli/npm/resolution.rs

```diff
@@ -0,0 +1,310 @@
+// Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.
+
+use std::collections::HashSet;
+use std::sync::Arc;
+
+use deno_core::error::AnyError;
+use deno_core::parking_lot::Mutex;
+use deno_core::parking_lot::RwLock;
+use deno_core::TaskQueue;
+use deno_lockfile::NpmPackageDependencyLockfileInfo;
+use deno_lockfile::NpmPackageLockfileInfo;
+use deno_npm::registry::NpmPackageInfo;
+use deno_npm::resolution::NpmPackagesPartitioned;
+use deno_npm::resolution::NpmResolutionSnapshot;
+use deno_npm::NpmPackageCacheFolderId;
+use deno_npm::NpmPackageId;
+use deno_npm::NpmResolutionPackage;
+use deno_semver::npm::NpmPackageNv;
+use deno_semver::npm::NpmPackageNvReference;
+use deno_semver::npm::NpmPackageReq;
+use deno_semver::npm::NpmPackageReqReference;
+
+use crate::args::Lockfile;
+
+use super::registry::NpmRegistry;
+
+/// Handles updating and storing npm resolution in memory.
+///
+/// This does not interact with the file system.
+#[derive(Clone)]
+pub struct NpmResolution(Arc<NpmResolutionInner>);
```

Comment:
I need a bit of explanation here - I thought that resolution would be abstracted to a separate crate. Does this implementation have some specific relation to the CLI?

Reply:
I'll update the description. It's specific to the CLI because it allows updating the resolution concurrently without blocking anything else and it combines stuff like the lockfile.